File: 695-max-area-of-island/695-max-area-of-island.java
class Solution {
    public int maxAreaOfIsland(int[][] grid) {

        int max=0;
        for (int i=0;i<grid.length;i++){
            for (int j=0 ;j<grid[0].length; j++){
                if(grid[i][j]==1){
                    int area=iarea(grid,i,j);
                    if(area >max)
                        max=area;
                }
            }
        }
        return max;
    }
    public static int iarea(int [][]grid,int i,int j){
        if(i<grid.length && j<grid[0].length && i>=0 &&  j>=0 && grid[i][j]==1){
            grid[i] [j]   =0;
            int up=iarea(grid,i-1,j);
            int down=iarea(grid, i+1,j);
            int left=iarea(grid, i,j-1);
            int right=iarea(grid,i,j+1);
            return 1+left+right+up+down;
        }
     return 0;       
    }
}

File: 695-max-area-of-island-in-python/695_max_area_of_island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        if not grid:
            return 0
        
        maxArea = 0
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    maxArea = max(maxArea, self.dfs(grid, i, j))
                    
        return maxArea
    
                    
    def dfs(self, grid, i, j):
        if i<0 or j<0 or i>=len(grid) or j>=len(grid[0]) or grid[i][j] != 1:
            return 0
        
        maxArea = 1
        grid[i][j] = '#'  # this will act as visited set
        maxArea += self.dfs(grid, i+1, j)
        maxArea += self.dfs(grid, i-1, j)
        maxArea += self.dfs(grid, i, j+1)
        maxArea += self.dfs(grid, i, j-1)
        
        return maxArea